modules/core/src/shaderlib/project/project-functions.js

```diff
@@ -13,11 +14,14 @@ import {addMetersToLngLat} from '@math.gl/web-mercator';
 // offset modes apply the y adjustment (unitsPerMeter2) when projecting z
 // LNG_LAT mode only use the linear scale.
 function lngLatZToWorldPosition(lngLatZ, viewport, offsetMode = false) {
-  const [longitude, latitude, z = 0] = lngLatZ;
-  const [X, Y] = viewport.projectFlat(lngLatZ);
-  const distanceScales = viewport.getDistanceScales(offsetMode && [longitude, latitude]);
-  const Z = z * distanceScales.unitsPerMeter[2];
-  return [X, Y, Z];
+  const p = viewport.projectPosition(lngLatZ);
+
+  if (offsetMode && viewport instanceof WebMercatorViewport) {
```

Comment:
Nit: It would be desirable to avoid `instanceof`. But probably best to save that for a rethink of the Viewports, so fine as is.